### ISSUE TITLE
Issues with commonmark.js benchmarks

### ISSUE THREAD

**Issue by dikmax** [dart-lang/markdown#134]
Some tests from commonmark.js benchmark suite took too much time to complete. Take a look at https://github.com/jgm/commonmark.js/blob/master/bench/samples/block-ref-flat.md and https://github.com/jgm/commonmark.js/blob/master/bench/samples/inline-links-nested.md.

**Comment by srawlins** [comment]
Thanks for the report, @dikmax, how long do they take? How long did you decide is "too long?" What, approximately, is your hardware?

Thanks!

**Comment by dikmax** [comment]
I run them on MacBook Pro (i7, early 2015).
`block-ref-flat` runs 10sec,  `inline-links-nested` runs 5 secs.
But it easy to make them parse forever just by adding some additional brackets. 

For example, this line:

```
 [[[[[[[[[[[[[[[[[[[[ this should not slow down anything ]]]]]]]]]]]]]]]]]]]]: q
```

parses in 6 seconds. 

But this, with just 3 additional brackets on each side,

```
[[[[[[[[[[[[[[[[[[[[[[[ this should not slow down anything ]]]]]]]]]]]]]]]]]]]]]]]: q
```

runs 1 minute.

**Comment by srawlins** [comment]
I'm confident this was fixed by #202.

The user time here shows that we're not quadratic, or whatever, anymore.

```dart
$ time dart bin/markdown.dart
[[[[[[[[[[[[[[[[[[[[ this should not slow down anything ]]]]]]]]]]]]]]]]]]]]: q
<p>[[[[[[[[[[[[[[[[[[[[ this should not slow down anything ]]]]]]]]]]]]]]]]]]]]: q</p>


real    0m2.025s
user    0m0.481s
sys     0m0.116s
srawlins-macbookair:dart-markdown srawlins$ time dart bin/markdown.dart
[[[[[[[[[[[[[[[[[[[[[[[ this should not slow down anything ]]]]]]]]]]]]]]]]]]]]]]]: q
<p>[[[[[[[[[[[[[[[[[[[[[[[ this should not slow down anything ]]]]]]]]]]]]]]]]]]]]]]]: q</p>


real    0m1.318s
user    0m0.495s
sys     0m0.119s
```
